R/derive_param_clinbenefit.R

```diff
@@ -0,0 +1,256 @@
+  reference_date <- assert_symbol(enquo(reference_date))
+  assert_data_frame(
+    dataset_adsl,
+    required_vars = vars(!!!subject_keys, !!reference_date)
+  )
+  assert_data_frame(
+    dataset,
+    required_vars = vars(!!!by_vars, PARAMCD, AVALC, ADT)
+  )
+
+  filter_source <- assert_filter_cond(
+    enquo(filter_source),
+    optional = FALSE
+  )
+
+  assert_vars(subject_keys)
+  assert_s3_class(source_resp, "date_source")
+  assert_s3_class(source_pd, "date_source")
+  assert_list_of(source_datasets, "data.frame")
+
+  source_names <- names(source_datasets)
+  if (!all(c(source_pd$dataset_name, source_resp$dataset_name) %in% source_names)) {
+    abort(
+      paste0(
+        "The dataset names specified for `source_pd` and `source_resp` must be \n",
+        "included in the list specified for the `source_datasets` parameter.\n",
+        "Following names were provided by `source_datasets`:\n",
+        enumerate(source_names, quote_fun = squote)
+      )
+    )
+  }
+
+  ref_start_window <- assert_integer_scalar(ref_start_window)
+  assert_varval_list(set_values_to, accept_expr = TRUE, optional = TRUE)
+  assert_param_does_not_exist(dataset, quo_get_expr(set_values_to$PARAMCD))
+
+  # ADSL variables
+
+  adsl_vars <- vars(
+    !!!subject_keys,
+    !!reference_date
+  )
+
+  adsl <- dataset_adsl %>%
+    select(!!!adsl_vars)
+
+  # Get PD date and objective response date
+
+  pd_data <- source_datasets[[source_pd$dataset_name]] %>%
+    filter_if(source_pd$filter) %>%
+    select(!!!subject_keys, !!!by_vars, !!source_pd$date) %>%
+    rename(temp_pd = !!source_pd$date)
+
+  rsp_data <- source_datasets[[source_resp$dataset_name]] %>%
+    filter_if(source_resp$filter) %>%
+    select(!!!subject_keys, !!!by_vars, !!source_resp$date) %>%
+    rename(temp_rs = !!source_resp$date)
+
+  # Look for valid non-PD measurements after window from reference date
+  ovr_data <- filter_pd(
+    dataset = dataset,
+    filter = !!filter_source,
+    source_pd = source_pd,
+    source_datasets = source_datasets
+  )
+
+  ovr_data <- ovr_data %>%
+    left_join(
+      adsl,
+      by = vars2chr(subject_keys)
+    ) %>%
+    left_join(
+      pd_data,
+      by = vars2chr(subject_keys)
+    ) %>%
+    filter(
+      !(AVALC %in% c("NA", "NE", "ND")) & !is.na(AVALC) &
+        ADT >= !!reference_date + days(ref_start_window)
+    ) %>%
+    # Use this approach only for patients that are not already responders
+    anti_join(
+      .,
+      rsp_data,
+      by = vars2chr(subject_keys)
+    ) %>%
+    filter_extreme(
+      order = vars(ADT),
+      by_vars = vars(!!!subject_keys, !!!by_vars),
+      mode = "first",
+      check_type = "none"
+    ) %>%
+    select(!!!subject_keys, !!!by_vars, ADT)
+
+  rsp_data <- rsp_data %>%
+    rename(ADT = temp_rs)
+
+  new_param <- bind_rows(ovr_data, rsp_data) %>%
+    right_join(adsl, by = vars2chr(subject_keys)) %>%
+    mutate(AVALC = if_else(!is.na(ADT), "Y", "N")) %>%
```

Comment:
`AVAL` is not set. I would provide `aval_fun` parameter like in https://github.com/pharmaverse/admiralonco/pull/44/.